src/keldysh_dlr.jl

```diff
@@ -0,0 +1,248 @@
+
+"""
+Extension of Keldysh.jl defining imaginary time Green's functions
+represented using the Discrete Lehmann Representation as implemented
+in Lehmann.jl.
+
+# Exports
+$(EXPORTS)
+"""
+module keldysh_dlr
+
+using DocStringExtensions
+
+using Lehmann; le = Lehmann
+
+using Keldysh; kd = Keldysh
+using Keldysh: AbstractTimeGrid
+using Keldysh: ImaginaryContour, TimeGridPoint, PeriodicStorage, GFSignEnum, BranchPoint
+
+import QInchworm.utility: ph_conj
+
+export DLRImaginaryTimeGrid, DLRImaginaryTimeGF
+export ph_conj
+
+#
+# DLRImaginaryTimeGrid
+#
+
+"""
+    $(TYPEDEF)
+
+Wrapper around Lehmann.jl describing a Discrete Lehmann Representation imaginary time grid
+conforming to the interface of Keldysh.jl TimeGrids.
+
+# Fields
+$(TYPEDFIELDS)
+"""
+struct DLRImaginaryTimeGrid <: AbstractTimeGrid
+    contour::ImaginaryContour
+    points::Vector{TimeGridPoint}
+    branch_bounds::NTuple{1, Pair{TimeGridPoint, TimeGridPoint}}
+    ntau::Int
+    dlr::le.DLRGrid
+
+    function DLRImaginaryTimeGrid(c::ImaginaryContour, dlr::le.DLRGrid)
+        points::Vector{TimeGridPoint} = []
+        τ_branch = c.branches[1]
+        @assert τ_branch.domain == kd.imaginary_branch
+        for (idx, τ) in enumerate(dlr.τ)
+            point = TimeGridPoint(1, idx, τ_branch(τ/dlr.β))
+            push!(points, point)
+        end
+        τ_0 = TimeGridPoint(-1, -1, τ_branch(0.))
+        τ_β = TimeGridPoint(-1, -1, τ_branch(1.))
+
+        branch_bounds = ( Pair(τ_0, τ_β), )
+        ntau = length(dlr.τ)
+        return new(c, points, branch_bounds, ntau, dlr)
+    end
+end
+
+Base.:isequal(A::T, B::T) where T <: DLRImaginaryTimeGrid = all(A.points .== B.points)
+
+"""
+    $(TYPEDEF)
+
+Wrapper around Lehmann.jl describing a Discrete Lehmann Representation imaginary time Green's
+function conforming to the interface of Keldysh.jl AbstractTimeGF.
+
+# Fields
+$(TYPEDFIELDS)
+"""
+struct DLRImaginaryTimeGF{T, scalar} <: AbstractTimeGF{T, scalar}
+    grid::DLRImaginaryTimeGrid
+    mat::PeriodicStorage{T,scalar}
+    ξ::GFSignEnum
+end
+
+"""
+    $(TYPEDSIGNATURES)
+
+Make a [`DLRImaginaryTimeGF`](@ref) from a [`DLRImaginaryTimeGrid`](@ref)
+following the API of Keldysh.ImaginarTimeGF.
+
+"""
+function DLRImaginaryTimeGF(::Type{T},
+                            grid::DLRImaginaryTimeGrid,
+                            norb=1,
+                            ξ::GFSignEnum=fermionic,
+                            scalar=false) where T <: Number
+    ntau = grid.ntau
+    mat = PeriodicStorage(T, ntau, norb, scalar)
+    return DLRImaginaryTimeGF(grid, mat, ξ)
+end
+
+DLRImaginaryTimeGF(grid::DLRImaginaryTimeGrid,
+                   norb=1,
+                   ξ::GFSignEnum=fermionic,
+                   scalar=false) = DLRImaginaryTimeGF(ComplexF64, grid, norb, ξ, scalar)
+
+norbitals(G::DLRImaginaryTimeGF) = G.mat.norb
+
+#
+# Arithmetic operations
+#
+
+function Base.:+(A::T, B::T) where T <: DLRImaginaryTimeGF
+    @assert A.grid == B.grid
```

Comment:
`A.ξ` and `B.ξ` must agree as well. (same in `Base.:-`)

Reply:
Agreed. I have added the same behaviour as in `isapprox` see above.